cmd/doctor.go

```diff
@@ -0,0 +1,191 @@
+package cmd
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"strings"
+
+	"github.com/exercism/cli/api"
+	"github.com/exercism/cli/config"
+	"github.com/exercism/cli/workspace"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+)
+
+// doctorCmd reports workspace cleanup tasks.
+var doctorCmd = &cobra.Command{
+	Use:     "doctor",
+	Aliases: []string{"doc"},
+	Short:   "Doctor reports workspace cleanup tasks.",
+	Long: `Doctor reports workspace cleanup tasks.
+
+	Use --fixup to run reported tasks.
+`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		cfg := config.NewConfig()
+
+		usrCfg := viper.New()
+		usrCfg.AddConfigPath(cfg.Dir)
+		usrCfg.SetConfigName("user")
+		usrCfg.SetConfigType("json")
+		// Ignore error. If the file doesn't exist, that is fine.
+		_ = usrCfg.ReadInConfig()
+		cfg.UserViperConfig = usrCfg
+
+		v := viper.New()
+		v.AddConfigPath(cfg.Dir)
+		v.SetConfigName("cli")
+		v.SetConfigType("json")
+		// Ignore error. If the file doesn't exist, that is fine.
+		_ = v.ReadInConfig()
+
+		return runDoctor(cfg, cmd.Flags())
+	},
+}
+
+func runDoctor(cfg config.Config, flags *pflag.FlagSet) error {
+	usrCfg := cfg.UserViperConfig
+	if usrCfg.GetString("workspace") == "" {
+		return fmt.Errorf(msgRerunConfigure, BinaryName)
+	}
+
+	fixup, err := flags.GetBool("fixup")
+	if err != nil {
+		return err
+	}
+	if fixup {
+		if err := runFixup(cfg); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func runFixup(cfg config.Config) error {
+	usrCfg := cfg.UserViperConfig
+
+	ws, err := workspace.New(usrCfg.GetString("workspace"))
+	if err != nil {
+		return err
+	}
+
+	exercises, err := ws.PotentialExercises()
+	if err != nil {
+		return err
+	}
+
+	if err := fixupMetadata(exercises, cfg); err != nil {
+		return err
+	}
+
+	// TODO: add numeric suffix dir cleanup #699
+
+	return nil
+}
+
+func fixupMetadata(exercises []workspace.Exercise, cfg config.Config) error {
+	for _, exercise := range exercises {
+		if _, err := exercise.MigrateLegacyMetadataFile(); err != nil {
+			return err
+		}
+		if ok, _ := exercise.HasMetadata(); !ok {
+			if err := downloadMetadata(exercise, cfg); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// TODO: extract this and cmd/download into download service?
+// copied from download#runDownload
+func downloadMetadata(exercise workspace.Exercise, cfg config.Config) error {
```

Comment:
unexported functions are available in the entire `cmd` package. We don't need to copy it, just extract it.

Reply:
Right. I wasn't suggesting that duplicating the same function. I was trying to point out that [runDownload](https://github.com/exercism/cli/blob/0cd96e0c880e5df9f9a484c85f75512e20c8f7a0/cmd/download.go#L51-L169) currently has duplicate behavior that we're going to want to use in `doctor` but we can't use the existing function. It seems that the logic for getting a `downloadPayload` should be extracted outside of the `cmd` package (given that multiple cmd's use it) with separate functions for writing metadata and solution files. I'm wondering if https://github.com/exercism/cli/issues/719 is related. I think this needs understood before proceeding with this PR.

Reply:
I'm not sure that it should live outside the cmd package, if only commands need it.

The code in api is very generic, whereas this is pretty specific.

I'd stick the extracted logic in `cmd/cmd.go` for now.

Reply:
Only the commands need it but it seems more like a service than an actual command. Everything in the `cmd` package is currently a command, so either the download payload stuff goes in `cmd/cmd` as a tacit home for shared command functionality or it could be put into a new 'helper/util' file. Maybe that's fine. That was my thinking anyway.